cmd/necoperf-cli/cmd/profile.go

```diff
@@ -0,0 +1,98 @@
+package cmd
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/cybozu-go/necoperf/internal/client"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+)
+
+var config struct {
+	namespace     string
+	podName       string
+	outputDir     string
+	containerName string
+	necoperfNS    string
+	timeout       time.Duration
+}
+
+func isPodReady(pod *corev1.Pod) bool {
+	for _, cond := range pod.Status.Conditions {
+		if cond.Type != corev1.PodReady {
+			continue
+		}
+		return cond.Status == corev1.ConditionTrue
+	}
+	return false
+}
+
+func NewProfileCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "profile",
+		Short: "Exec perf profile of the target container",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			config.podName = args[0]
+			handler := slog.NewTextHandler(os.Stderr, nil)
+			logger := slog.New(handler)
+
+			client, err := client.New(logger, config.timeout)
+			if err != nil {
+				return err
+			}
+			if err := client.SetupDiscovery(); err != nil {
+				return err
+			}
+
+			ctx := cmd.Context()
+			pod, err := client.Discovery.GetPod(ctx, config.namespace, config.podName)
+			if err != nil {
+				return err
+			}
+			if !isPodReady(pod) {
+				return fmt.Errorf("pod %s is not ready", pod.Name)
+			}
+			containerID, err := client.Discovery.GetContainerID(pod, config.containerName)
+			if err != nil {
+				return err
+			}
+			logger.Info("get container id", "podName", config.podName, "containerID", containerID)
+
+			pods, err := client.Discovery.GetPodList(ctx, config.necoperfNS)
+			if err != nil {
+				return err
+			}
+			addr, err := client.Discovery.DiscoveryServerAddr(pods, pod.Status.HostIP)
+			if err != nil {
+				return err
+			}
+			err = client.SetupGrpcClient(addr)
+			if err != nil {
+				return err
+			}
+			logger.Info("connect grpc server", "addr", addr)
+
+			err = client.Profile(ctx, config.podName, containerID, config.outputDir)
+			if err != nil {
+				return err
+			}
+			logger.Info("profile is finished", "output directory", filepath.Join(config.outputDir, config.podName+".script"))
+
+			return nil
+		},
+		SilenceUsage:  true,
```

Comment:
As you say, it is not user-friendly to not display a help message when incorrect arguments are passed.

I was not sure how this should be implemented, so may I ask you to elaborate on this?
>It would be good to set cmd.SilenceUsage = true at the head of RunE.

Reply:
@zeroalphat 
The following code would be helpful.

https://github.com/cybozu-go/sabakan/blob/v2.13.2/pkg/sabactl/cmd/root.go#L62

`PersistentPreRunE` is executed only when the arguments fill the requirements of `Args` and `Flags`.
`sabactl` starts its execution with `SilenceUsage == false (default)`, verifies its arguments, and then run its `PersistentPreRunE` and `RunE` with `SilenceUsage` changing to `true`.
So a run-time error in `RunE` does not display the help message.

Reply:
@morimoto-cybozu 
Thank you for explaining in detail.
It helped me to understand how to implement it.